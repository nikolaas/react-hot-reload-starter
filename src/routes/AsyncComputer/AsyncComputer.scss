.AsyncComputer__title {
    display: flex;
    align-items: center;
}

.AsyncComputer__icon {
    background: url('../../assets/images/caclulator.png') no-repeat center center;
    display: inline-block;
    width: 32px;
    height: 32px;
}

.AsyncComputer__calculate-icon {
    width: 16px;
    height: 16px;
    margin-right: 1em;
    fill: #000;
}

.AsyncComputer__calculate {
    display: flex;
    justify-content: center;
    align-items: center;
    width: 10em;

    &:hover {
        .AsyncComputer__calculate-icon {
            fill: #f00
        }
    }
}
