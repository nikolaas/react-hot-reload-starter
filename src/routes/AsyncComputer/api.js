export default {
    sum: async (a, b) => {
        /* eslint-disable no-console, no-unused-vars */
        console.log(`Sum request ${a} + ${b}`);
        // emulate request to backend
        return new Promise((resolve, reject) => {
            setTimeout(() => {
                resolve(a + b);
                // reject(new Error('Some backend error'));
            }, 500);
        });
    },
};
