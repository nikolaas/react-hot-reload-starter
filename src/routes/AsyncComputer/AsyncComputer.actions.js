import api from './api';

export const CHANGE_ARGUMENT = 'change_argument';
export const COMPUTING_STARTED = 'computing_started';
export const COMPUTING_SUCCESS = 'computing_success';
export const COMPUTING_ERROR = 'computing_error';

export const changeArgument = (name, value) => {
    return { type: CHANGE_ARGUMENT, payload: { name, value } };
};

export const startComputing = () => {
    return { type: COMPUTING_STARTED };
};

export const successComputing = result => {
    return { type: COMPUTING_SUCCESS, payload: result };
};

export const errorComputing = error => {
    return { type: COMPUTING_ERROR, payload: error };
};

export const sum = () => async (dispatch, getState) => {
    const state = getState().asyncComputer;
    const { a, b } = state;
    dispatch(startComputing());
    try {
        const result = await api.sum(a, b);
        dispatch(successComputing(result));
    } catch (error) {
        dispatch(errorComputing(String(error)));
    }
};
