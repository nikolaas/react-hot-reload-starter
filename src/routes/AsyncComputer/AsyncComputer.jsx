import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { BaseLayout } from '../../components/BaseLayout';
import { Button } from '../../components/Button';
import { Article } from '../../components/Article';
import { changeArgument, sum } from './AsyncComputer.actions';
import CalculateIcon from '../../assets/images/calculate.dynamic.svg';
import './AsyncComputer.scss';

export class RawAsyncComputer extends React.Component {

    static propTypes = {
        a: PropTypes.number,
        b: PropTypes.number,
        computing: PropTypes.bool,
        result: PropTypes.number,
        error: PropTypes.string,
        onArgumentChange: PropTypes.func.isRequired,
        onSum: PropTypes.func.isRequired,
    };

    static defaultProps = {
        a: null,
        b: null,
        computing: false,
        result: null,
        error: null,
    };

    handleInputChange = event => {
        const { name, value } = event.target;
        const parsedValue = parseInt(value, 10) || null;
        this.props.onArgumentChange(name, parsedValue);
    };

    handleResultClick = async () => {
        this.props.onSum();
    };

    render() {
        const { a, b, computing, error, result } = this.props;
        const title = <span className="AsyncComputer__title"><i className="AsyncComputer__icon"/>Async Computer</span>;
        return (
            <BaseLayout>
                <Article title={title}>
                    <p>
                        <label htmlFor="a">
                            {'a = '}
                            <input
                                type="text"
                                id="a"
                                name="a"
                                value={a == null ? '' : a}
                                disabled={computing}
                                onChange={this.handleInputChange}
                            />
                        </label>
                    </p>
                    <p>
                        <label htmlFor="b">
                            {'b = '}
                            <input
                                type="text"
                                id="b"
                                name="b"
                                value={b == null ? '' : b}
                                disabled={computing}
                                onChange={this.handleInputChange}
                            />
                        </label>
                    </p>
                    <Button className="AsyncComputer__calculate" disabled={computing} onClick={this.handleResultClick}>
                        <CalculateIcon className="AsyncComputer__calculate-icon"/> Sum
                    </Button>
                    {
                        computing &&
                        <p className="AsyncComputer__result">Computing...</p>
                    }
                    {
                        !computing && error &&
                        <p className="AsyncComputer__result">Error: <span style={{ color: '#f00' }}>{error}</span></p>
                    }
                    {
                        !computing && result != null &&
                        <p className="AsyncComputer__result">Sum: <span style={{ color: '#f00' }}>{result}</span></p>
                    }
                </Article>
            </BaseLayout>
        );
    }

}

const mapStateToProps = state => {
    const { asyncComputer } = state;
    const { a, b, computing, result, error } = asyncComputer;
    return { a, b, computing, result, error };
};

const mapDispatchToProps = {
    onArgumentChange: changeArgument,
    onSum: sum,
};

export default connect(mapStateToProps, mapDispatchToProps)(RawAsyncComputer);
