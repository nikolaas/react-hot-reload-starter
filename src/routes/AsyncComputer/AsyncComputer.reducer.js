import { CHANGE_ARGUMENT, COMPUTING_STARTED, COMPUTING_SUCCESS, COMPUTING_ERROR } from './AsyncComputer.actions';

const initialState = {
    a: null,
    b: null,
    computing: false,
    result: null,
    error: null,
};

export default (state = initialState, action = {}) => {
    switch (action.type) {
        case CHANGE_ARGUMENT: {
            return { ...state, [action.payload.name]: action.payload.value };
        }
        case COMPUTING_STARTED: {
            return { ...state, computing: true, result: null, error: null };
        }
        case COMPUTING_SUCCESS: {
            return { ...state, computing: false, result: action.payload, error: null };
        }
        case COMPUTING_ERROR: {
            return { ...state, computing: false, result: null, error: String(action.payload) };
        }
        default: {
            return state;
        }
    }
};
