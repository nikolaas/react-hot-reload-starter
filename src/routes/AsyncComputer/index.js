import { registerAsyncReducer } from '../../store';
import AsyncComputer from './AsyncComputer';
import reducer from './AsyncComputer.reducer';

registerAsyncReducer('asyncComputer', reducer);

export default AsyncComputer;
