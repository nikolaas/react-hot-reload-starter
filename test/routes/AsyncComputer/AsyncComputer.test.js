import React from 'react';
import { shallow } from 'enzyme';
import { RawAsyncComputer } from '../../../src/routes/AsyncComputer/AsyncComputer';

describe('Sync counter tests', () => {
    it('should render arguments values in inputs when arguments is specified', () => {
        const component = shallow(<RawAsyncComputer onArgumentChange={jest.fn()} onSum={jest.fn()} a={1} b={2}/>);
        expect(component.find('#a').prop('value')).toMatchSnapshot();
        expect(component.find('#b').prop('value')).toMatchSnapshot();
    });

    it('should render "computing..." message when computing in progress', () => {
        const component2 = shallow(<RawAsyncComputer onArgumentChange={jest.fn()} onSum={jest.fn()} computing/>);
        expect(component2.find('.AsyncComputer__result').text()).toMatchSnapshot();
    });

    it('should render result of computing when computing finished successfully', () => {
        const component = shallow(<RawAsyncComputer onArgumentChange={jest.fn()} onSum={jest.fn()} result={3}/>);
        expect(component.find('.AsyncComputer__result').text()).toMatchSnapshot();
    });

    it('should render error of computing when computing finished with error', () => {
        const component = shallow(<RawAsyncComputer onArgumentChange={jest.fn()} onSum={jest.fn()} error="error"/>);
        expect(component.find('.AsyncComputer__result').text()).toMatchSnapshot();
    });

    it('should trigger onArgumentChange when some argument\'s values was changed', () => {
        const argumentChangeSpy = jest.fn();
        const component = shallow(<RawAsyncComputer onArgumentChange={argumentChangeSpy} onSum={jest.fn()}/>);

        component.find('#a').simulate('change', { target: { name: 'a', value: '1' } });
        expect(argumentChangeSpy).toHaveBeenLastCalledWith('a', 1);

        component.find('#b').simulate('change', { target: { name: 'b', value: '2' } });
        expect(argumentChangeSpy).toHaveBeenLastCalledWith('b', 2);
    });

    it('should trigger onSum when sum button was clicked', () => {
        const sumSpy = jest.fn();
        const component = shallow(<RawAsyncComputer onArgumentChange={jest.fn()} onSum={sumSpy}/>);

        component.find('Button').simulate('click');

        expect(sumSpy).toHaveBeenCalled();
    });
});
