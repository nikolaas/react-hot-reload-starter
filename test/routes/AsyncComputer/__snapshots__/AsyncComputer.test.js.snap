// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sync counter tests should render "computing..." message when computing in progress 1`] = `"Computing..."`;

exports[`Sync counter tests should render arguments values in inputs when arguments is specified 1`] = `1`;

exports[`Sync counter tests should render arguments values in inputs when arguments is specified 2`] = `2`;

exports[`Sync counter tests should render error of computing when computing finished with error 1`] = `"Error: error"`;

exports[`Sync counter tests should render result of computing when computing finished successfully 1`] = `"Sum: 3"`;
